src/Transformer/DetermineHeroImages.php

```diff
@@ -0,0 +1,173 @@
+ *
+ * @package AmpProject\AmpWP
+ * @since   2.1
+ * @internal
+ */
+final class DetermineHeroImages implements Transformer {
+
+	/**
+	 * XPath query to find the existing hero images (elements with data/hero attribute).
+	 *
+	 * @var string
+	 */
+	const EXISTING_HERO_IMAGES_XPATH_QUERY = './/*[@data-hero]';
+
+	/**
+	 * XPath query to find the site icon.
+	 *
+	 * @var string
+	 */
+	const SITE_ICON_XPATH_QUERY = ".//a[ contains( concat( ' ', normalize-space( @class ), ' ' ), ' custom-logo-link ' ) ]//*[ contains( concat( ' ', normalize-space( @class ), ' ' ), ' custom-logo ' ) ][ not( @data-hero ) ]";
+
+	/**
+	 * XPath query to find the featured image.
+	 *
+	 * @var string
+	 */
+	const FEATURED_IMAGE_XPATH_QUERY = ".//*[ contains( concat( ' ', normalize-space( @class ), ' ' ), ' wp-post-image ' ) ][ not( @data-hero ) ]";
+
+	/**
+	 * XPath query to find the cover blocks.
+	 *
+	 * @var string
+	 */
+	const COVER_BLOCKS_XPATH_QUERY = ".//*[ contains( concat( ' ', normalize-space( @class ), ' ' ), ' wp-block-cover ' ) ][ not( @data-hero ) ]";
+
+	/**
+	 * Apply transformations to the provided DOM document.
+	 *
+	 * @param Document        $document DOM document to apply the
+	 *                                  transformations to.
+	 * @param ErrorCollection $errors   Collection of errors that are collected
+	 *                                  during transformation.
+	 * @return void
+	 */
+	public function transform( Document $document, ErrorCollection $errors ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		$existing_hero_images_count = $document->xpath->query(
+			self::EXISTING_HERO_IMAGES_XPATH_QUERY,
+			$document->body
+		)->length;
+
+		$available_hero_image_slots = max(
+			PreloadHeroImage::DATA_HERO_MAX - $existing_hero_images_count,
+			0
+		);
+
+		$hero_image_elements = [];
+
+		if ( count( $hero_image_elements ) < $available_hero_image_slots ) {
+			$site_icon = $this->get_site_icon( $document );
+			if ( null !== $site_icon ) {
+				$hero_image_elements[] = $site_icon;
+			}
+		}
+
+		if ( count( $hero_image_elements ) < $available_hero_image_slots ) {
+			$featured_image = $this->get_featured_image( $document );
+			if ( null !== $featured_image ) {
+				$hero_image_elements[] = $featured_image;
+			}
+		}
+
+		if ( count( $hero_image_elements ) < $available_hero_image_slots ) {
+			$hero_image_elements = array_merge(
+				$hero_image_elements,
+				array_filter(
+					$this->get_cover_blocks( $document )
+				)
+			);
+		}
+
+		$this->add_data_hero_attribute(
+			array_slice( $hero_image_elements, 0, $available_hero_image_slots )
+		);
+	}
+
+	/**
+	 * Retrieve the element that represents the site icon.
+	 *
+	 * @param Document $document Document to retrieve the site icon from.
+	 * @return DOMElement|null Element that represents the site icon, or null
+	 *                         if not found.
+	 */
+	private function get_site_icon( Document $document ) {
```

Comment:
This should be called `custom_logo` instead of `site_icon`, since in WordPress the site icon is specifically the favicon.